doc/api18970.md

```diff
@@ -0,0 +1,570 @@
+invoked by a thread it is possible for that thread to be suspended as a
+consequence of a higher-priority thread being made ready.  Whether the
+suspension actually occurs depends on the operation associated with the
+reschedule point and the relative priorities of the invoking thread and
+the head of the ready queue.
+
+Note that in the case of timeslicing, or reschedule points executed from
+interrupts, any thread may be suspended in any function.
+
+Functions that are not **reschedule** may be invoked from either thread
+or interrupt context.
+
+Functions that are **reschedule** may be invoked from thread context.
+
+Functions that are **reschedule** but not **sleep** may be invoked from
+interrupt context.
+
+### sleep
+
+The sleep attribute is used on a function that can cause the invoking
+thread to sleep.
+
+Functions that are **sleep** are implicitly **reschedule**.
+
+Functions that are **sleep** may be invoked from thread context.
+
+Functions that are **sleep** may be invoked from interrupt context if
+and only if invoked in **no-wait** mode.
+
+#### Commentary
+
+This attribute is of relevance specifically when considering application
+that use only cooperative threads, because the kernel will not replace a
+running cooperative thread at a reschedule point unless that thread has
+explicitly invoked an operation that caused it to sleep.
+
+Note that this attribute does not imply the function will sleep
+unconditionally, but that the state of the system may require the
+invoking thread to suspend, wait, or invoke `k_yield()` before it can
+complete its operation.  This behavior may be mediated by **no-wait**.
+
+### no-wait
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+Functions with this attribute may be invoked from interrupt context when
+the parameter selects the no-wait path.
+
+Functions that are **no-wait** functions are implicitly **sleep**.
+
+#### Commentary
+
+The paradigmatic case of a no-wait function is a function that takes a
+timeout, to which `K_NO_WAIT` can be passed.  The semantics of this
+special timeout value are to execute the function's operation as long as
+it can be completed immediately, and to return an error code rather than
+sleep if it cannot.
+
+It is use of the no-wait feature that allows functions like
+`k_sem_take()` to be invoked from ISRs, since it is not permitted to
+sleep in interrupt context.
+
+**NOTE** That a function has a no-wait path does not imply that taking
+that path guarantees the function is synchronous.
+
+### isr-ok
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context.  The function will use `k_is_in_isr()` to
+detect its calling context and force an execution path that will not
+cause the invoking thread to sleep.
+
+This attribute is intended for **sleep** functions that may be
```

Comment:
```suggestion
This attribute is useful for **sleep** functions that may be
```

Just a suggestion, since again `k_busy_wait()` is isr-ok but not sleep.

In general I think it's a bit cleaner if the definition of the term `isr-ok` doesn't assume that functions so marked are `sleep`, even if that's the primary use case.

Reply:
I've taken the previous suggestion to add "If necessary" regarding `k_is_in_isr()`, but for now I'm going to stick with "intended" because that is its purpose.  "intended" shouldn't imply "reserved for use by".

At this time the only guidance I have for what makes something callable from an interrupt is that "it doesn't sleep".  Nobody's explained what "sleep" means, and my attempts to define it rigorously have led me deep into rabbit warrens so I'm just gonna hand-wave it.

People could tag non-sleep functions with `isr-ok`, but I'm very concerned about a slippery slope where people say "Hey, `k_cycle_get()` (or `k_ms_to_ticks_floor32()` or `crc16()`) isn't marked `isr-ok` so I can't use it in an interrupt?" and we end up explicitly marking a huge number of functions that have nothing to do with thread sleep behavior as `isr-ok`.

If we want complete explicit attribute sets for every API function I'd rather see some tooling that synthesizes derived attributes based on the presence/absence of indicator attributes and populates those in the generated documentation.  Anything that isn't marked `sleeps` should be assumed to be callable from an interrupt.  (Whether doing so is a good idea is a different issue.)

Reply:
> I'm going to stick with "intended"

ACK

> I'm very concerned about a slippery slope where [...] we end up explicitly marking a huge number of functions that have nothing to do with thread sleep behavior as `isr-ok`.
>
> If we want complete explicit attribute sets for every API function I'd rather see some tooling that synthesizes derived attributes based on the presence/absence of indicator attributes and populates those in the generated documentation. 

OK. Perhaps I'll understand this concern better when you have time to show how you plan on using these terms in a non-draft context to mark up existing APIs in RST documentation.